plugins/inputs/gtfs/gtfs.go

```diff
@@ -0,0 +1,305 @@
+	VehiclePositionsURL string `toml:"vehicle_positions_url"`
+	TripUpdatesURL      string `toml:"trip_updates_url"`
+	ServiceAlertsURL    string `toml:"service_alerts_url"`
+
+	Username string            `toml:"username"`
+	Password string            `toml:"password"`
+	Timeout  internal.Duration `toml:"timeout"`
+
+	tls.ClientConfig
+
+	client              *http.Client
+	vehiclePositionsReq *http.Request
+	tripUpdatesReq      *http.Request
+	serviceAlertsReq    *http.Request
+}
+
+func (s *GTFS) Description() string {
+	return "GTFS-realtime data"
+}
+
+func (s *GTFS) SampleConfig() string {
+	return `
+  ## API Key
+  # key = "${GTFS_API_KEY}"
+  ## URL for fetching vehicle positions
+  # vehicle_positions_url = "https://host.test/VehiclePositions.pb"
+  ## URL for fetching vehicle positions
+  # trip_updates_url = "https://host.test/TripUpdates.pb"
+  ## URL for fetching vehicle positions
+  # service_alerts_url = "https://host.test/ServiceAlerts.pb"
+
+  ## Optional HTTP Basic Auth Credentials
+  # username = "username"
+  # password = "pa$$word"
+
+  ## Optional TLS Config
+  # tls_ca = "/etc/telegraf/ca.pem"
+  # tls_cert = "/etc/telegraf/cert.pem"
+  # tls_key = "/etc/telegraf/key.pem"
+  ## Use TLS but skip chain & host verification
+  # insecure_skip_verify = false
+
+  ## Amount of testTimestamp allowed to complete the HTTP request
+  # timeout = "5s"
+`
+}
+
+func (g *GTFS) Init() error {
+	tlsCfg, err := g.ClientConfig.TLSConfig()
+	if err != nil {
+		return err
+	}
+
+	g.client = &http.Client{
+		Transport: &http.Transport{
+			TLSClientConfig: tlsCfg,
+			Proxy:           http.ProxyFromEnvironment,
+		},
+		Timeout: g.Timeout.Duration,
+	}
+
+	if g.VehiclePositionsURL == "" && g.TripUpdatesURL == "" && g.ServiceAlertsURL == "" {
+		return errNoSourceConfigured
+	}
+
+	if g.VehiclePositionsURL != "" {
+		r, err := g.newRequest(g.VehiclePositionsURL)
+		if err != nil {
+			return err
+		}
+		g.vehiclePositionsReq = r
+	}
+
+	if g.TripUpdatesURL != "" {
+		r, err := g.newRequest(g.TripUpdatesURL)
+		if err != nil {
+			return err
+		}
+		g.tripUpdatesReq = r
+	}
+
+	if g.ServiceAlertsURL != "" {
+		r, err := g.newRequest(g.ServiceAlertsURL)
+		if err != nil {
+			return err
+		}
+		g.serviceAlertsReq = r
+	}
+
+	return nil
+}
+
+func (g *GTFS) Gather(acc telegraf.Accumulator) error {
+	var (
+		work errgroup.Group
+		now  = time.Now()
+	)
+
+	if g.vehiclePositionsReq != nil {
+		work.Go(func() error { return g.gatherVehiclePositions(acc, now) })
+	}
+
+	if g.tripUpdatesReq != nil {
+		work.Go(func() error { return g.gatherTripUpdates(acc, now) })
+	}
+
+	if g.serviceAlertsReq != nil {
+		work.Go(func() error { return g.gatherServiceAlerts(acc, now) })
+	}
+
+	return work.Wait()
+}
+
+func (g *GTFS) gatherVehiclePositions(acc telegraf.Accumulator, t time.Time) error {
+	entities, err := g.gather(g.vehiclePositionsReq)
+	if err != nil {
+		return err
+	}
+
+	for _, entity := range entities {
+		if entity.Vehicle == nil {
+			continue
```

Comment:
Is this a common case or an error case. In case of the latter, please log a warning or an error here to make the user aware that metrics were dropped.

Reply:
I'm not aware of how common this is in the wild, but it is also not an error. Given the GTFS-realtime specification, it is entirely possible for a transit service to expose all three feed types (position, alerts and updates) in a single feed. This check allows the plugin to work gracefully in that case.

Reply:
The background of the question is that if you _expect_ that field to be there, we don't want to silently drop the received data but rather make the user aware that we threw away data. In this case I do expect a info or even warn message.
If an the other hand receiving multiple types of messages is a common thing, then we should not bombard the user with messages in the log...
I rely on your estimate how likely it is we drop data erroneously here... But maybe it is worth a comment for the next guy looking at the code.